src/scancode/cli.py

```diff
@@ -725,26 +725,20 @@ def _resource_logger(logfile_fd, resources):
     yield back the resources.
     """
     file_logger = ScanFileCache.log_file_path
-    for posix_path, rel_path in resources:
-        file_logger(logfile_fd, rel_path)
-        yield posix_path, rel_path
+    for resource in resources:
+        file_logger(logfile_fd, resource.rel_path)
+        yield resource
 
 
-def _scanit(paths, scanners, scans_cache_class, diag, timeout=DEFAULT_TIMEOUT, processes=1):
+def _scanit(resource, scanners, scans_cache_class, diag, timeout=DEFAULT_TIMEOUT, processes=1):
     """
     Run scans and cache results on disk. Return a tuple of (success, scanned relative
     path) where sucess is True on success, False on error. Note that this is really
     only a wrapper function used as an execution unit for parallel processing.
     """
-    abs_path, rel_path = paths
-    # always fetch infos and cache.
-    infos = OrderedDict()
-    infos['path'] = rel_path
-    infos.update(scan_infos(abs_path, diag=diag))
-
     success = True
     scans_cache = scans_cache_class()
-    is_cached = scans_cache.put_info(rel_path, infos)
+    is_cached = scans_cache.put_info(resource.rel_path, resource.infos)
```

Comment:
Here we put in the cache, but I think that at this stage, the `info` has already been fetched when iterating Resource before but was not cached then? IMHO we likely need to properly define this:
- when we first collect the `info` they should be cached then
- when we work on a file, check if has info or results cached and use these from the cache

I reckon that I am not 100% clear on what the current step-by-step process is!

Reply:
@pombredanne  a couple of questions,
 - Can a whole Resource object be cached?
 - Does a `Resource` object need to be cached? I mean can't we pass around the object as is and all the scan info can be accessed from it directly?

Reply:
One approach could be indeed to have the Resource object being something that is "cached" behind the scenes, or at least hides the caching logic
The caching logic seems simple but this is not entirely trivial to get something that is fast (and actually helping getting things to be faster) and required several iterations to get it right and working nicely on all OSes. For instance, an earlier version used a SQLite-backed cache and was a catastrophic choke point.

A Resource object (or at least its underlying data) needs to be cached if its attributes were computed from the actual file (e.g. checksums, attributes) otherwise these would need to be recomputed each time. For instance in this new approach, the "info" are collected as part of the resource iteration for the pre-scan plugins. This is done for now in a single process... AND once the pre-scan plugins have been processes in the stream, the actual scans are enqueued and handled asynchronously, sending their results back in the cache. 
The caching proper is keyed with the file "info" 

Resource cannot be all kept in memory as this would blow up size-wise (and is the main reason for the existence of a cache in the first place).

So all this is not 100% straightforward and I think we would best benefit from documenting the steps in the process and can then refactor this as needed.

Reply:
Ok, what I get from that is:
We need to add the caching logic for the redundant data such as checksums, attributes, etc. in the `Resource` object itself, thus when the `Resource` object's methods are called, they lookup the cache and return the data from there or recompute the data and store it in cache if not found. Am I getting this right?

Reply:
@pombredanne ping?

Reply:
@yashdsaraf you wrote
> We need to add the caching logic for the redundant data such as checksums, attributes, etc. in the Resource object itself, thus when the Resource object's methods are called, they lookup the cache and return the data from there or recompute the data and store it in cache if not found. Am I getting this right?

That sounds like a decent approach.